### DISCUSSION TITLE
Surgical Instrument Detection – Annotation Best Practices and YOLOv12n Inference Issue

### DISCUSSION THREAD

**Discussion by Tixtor710** [ultralytics/ultralytics#20670] [Q&A]
I’m building a **YOLOv12n-based surgical instrument detection model** to assist during real-time surgeries.

### Setup

- **Custom dataset:** [Roboflow Surgical Instrument Detection](https://universe.roboflow.com/computer-vision-gflxc/surgical-instrument-detection-cpjui)
- **Format:** Exported in **YOLOv12** format
- **Training:** YOLOv12n model, trained on **Kaggle** for **100 epochs**

Everything trained without issues. However, during **inference**, the model sometimes predicts **default COCO classes** — despite the dataset only including surgical instruments.

---

### Annotation Strategy

- Should I label each instrument separately (e.g., `scalpel`, `forceps`, `scissors`) or just use a general `"instrument"` class?
- How should I handle edge cases like:
  - Occlusion
  - Motion blur
  - Overlapping instruments
  - Blood-covered or partially visible tools

---

### Annotation Considerations

- What are **critical things to keep in mind** when labeling surgical frames?
- How do **lighting**, **glare**, and **reflections** affect detection performance?
- Should certain difficult cases be omitted or annotated differently?

---

### Inference Class Restriction

- How can I **ensure the YOLOv12n model only detects my custom classes** and does not include COCO defaults like `"person"` or `"car"`?
- Is there a **flag**, config, or **class list** that must be defined during inference?

---

Any input from others who’ve worked in medical vision or surgical detection would be greatly appreciated.  
Thanks in advance for your help!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Tixtor710, thank you for your detailed questions and for using Ultralytics 🚀! This is an automated response to help you get started—an Ultralytics engineer will follow up soon to provide more specific insights, especially regarding annotation best practices and medical vision considerations.

For new users, we recommend checking out the [Docs](https://docs.ultralytics.com/), which include many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides. Many topics around dataset preparation, annotation, and inference configuration are covered there and may answer some of your questions.

If your post involves a 🐛 Bug Report (such as the inference showing COCO classes), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)—including your inference command, sample predictions, and exported dataset YAML. This will help us debug and ensure your model is referencing only your custom classes.

If you have custom training ❓ questions, please share additional details like dataset snippets, annotation examples, and your training/inference logs. For annotation guidance, see our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community to connect with others working on similar projects:
- For live chat and quick support: [Discord](https://discord.com/invite/ultralytics) 🎧
- For longer discussions and sharing best practices: [Discourse](https://community.ultralytics.com/)
- To connect with other Ultralytics users: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

First, please ensure you are running the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This can resolve many unexpected issues:

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of the following up-to-date, verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks for sharing your project and questions! An Ultralytics engineer will review your discussion and provide customized advice soon.

**Reply by glenn-jocher** [reply]
Hi @Tixtor710, 

Thanks for sharing details about your surgical instrument detection project. I have some specific recommendations for your questions:

### Regarding COCO classes appearing during inference

This is happening because your model is likely still referencing the COCO classes. When using your custom model, make sure to specify your classes in the data YAML file:

```yaml
names:
  0: forceps
  1: scalpel
  2: scissors
  # Add all your surgical instrument classes
```

Then during inference, explicitly pass this class list:

```python
from ultralytics import YOLO
model = YOLO('your_trained_model.pt')
results = model('your_image.jpg', classes=[0, 1, 2])  # Only detect your surgical instruments
```

### Annotation strategy recommendations

1. **Use specific instrument classes** rather than a general "instrument" class - this provides more detailed information during surgery which could be critical.

2. **Edge cases handling:**
   - For occluded instruments: Label the visible part and maintain consistency
   - For motion blur: Include these in your dataset as they'll occur in real-time scenarios
   - For overlapping instruments: Label each instrument separately with accurate boundaries
   - For blood-covered tools: Include these variations for robustness

3. **Critical annotation considerations:**
   - Consistent bounding box placement (e.g., include handles consistently)
   - Account for different viewing angles and distances
   - Include reflective surfaces and glare conditions in your dataset

Based on our documentation for healthcare applications, I'd also suggest switching to YOLO11 models as they're more stable and perform better than YOLO12, especially for medical applications requiring reliability.

Would you be open to sharing a sample of your data YAML file and inference code so I can help troubleshoot the class detection issue more specifically?

**Reply by Tixtor710** [reply]
Thanks for taking a look! Yep, I’ve attached my inference script as a Markdown file, and the YAML file details are included above. Let me know if anything looks off or if you need a sample output from the predictions to debug further.
[YOLOv12n_Training_and_Inference.md](https://github.com/user-attachments/files/20301467/YOLOv12n_Training_and_Inference.md)

Yaml file:

```
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 9
names: ['Bovie pencil', 'Laparoscopic Grasper', 'Laparoscopic Maryland Dissector', 'Laparoscopic Trocar Cannula', 'Surgical Clamp', 'forceps', 'person-Bicycle-dog-cat', 'self-retaining retractor', 'sterile gauze']

roboflow:
  workspace: computer-vision-gflxc
  project: dataset-trial
  version: 1
  license: CC BY 4.0
  url: https://universe.roboflow.com/computer-vision-gflxc/dataset-trial/dataset/1
```

@glenn-jocher

**Reply by glenn-jocher** [reply]
Hi @Tixtor710,

Looking at your YAML file, I spotted a key issue that likely explains why your model is detecting COCO classes. The class name `'person-Bicycle-dog-cat'` is problematic - it contains common COCO class names that might be confusing your model.

I recommend:

1. **Clean up your class list** - Remove or rename that hybrid class to something specific to surgical instruments
2. **Switch to YOLO11** instead of YOLOv12n - As mentioned in our documentation, YOLO11 is more stable and recommended over YOLO12 which can be unstable during training, uses more memory, and runs 2-3x slower on CPU due to its attention layers

For class restriction during inference, add this to your code:

```python
# Explicitly filter to only your surgical classes
results = model(img, classes=[0, 1, 2, 3, 4, 5, 7, 8])  # Excluding the problematic class #6
```

For annotation best practices in medical contexts, our AI in healthcare documentation suggests consistent bounding box placement across similar instruments and including varied lighting conditions in your dataset. Surgical tools with reflective surfaces benefit from comprehensive training across different angles and lighting conditions.

Would you be able to share a few example prediction images showing the incorrect COCO classes being detected?
